miniapps/nurbs/nurbs_ex25p.cpp

```diff
@@ -0,0 +1,753 @@
+   PMLDiagMatrixCoefficient pml_c1_Im(cdim,detJ_inv_JT_J_Im, pml);
+   ScalarVectorProductCoefficient c1_Re(muinv,pml_c1_Re);
+   ScalarVectorProductCoefficient c1_Im(muinv,pml_c1_Im);
+   VectorRestrictedCoefficient restr_c1_Re(c1_Re,attrPML);
+   VectorRestrictedCoefficient restr_c1_Im(c1_Im,attrPML);
+
+   PMLDiagMatrixCoefficient pml_c2_Re(dim, detJ_JT_J_inv_Re,pml);
+   PMLDiagMatrixCoefficient pml_c2_Im(dim, detJ_JT_J_inv_Im,pml);
+   ScalarVectorProductCoefficient c2_Re(omeg,pml_c2_Re);
+   ScalarVectorProductCoefficient c2_Im(omeg,pml_c2_Im);
+   VectorRestrictedCoefficient restr_c2_Re(c2_Re,attrPML);
+   VectorRestrictedCoefficient restr_c2_Im(c2_Im,attrPML);
+
+   // Integrators inside the PML region
+   a.AddDomainIntegrator(new CurlCurlIntegrator(restr_c1_Re),
+                         new CurlCurlIntegrator(restr_c1_Im));
+   a.AddDomainIntegrator(new VectorFEMassIntegrator(restr_c2_Re),
+                         new VectorFEMassIntegrator(restr_c2_Im));
+
+   // 10. Assemble the bilinear form and the corresponding linear system,
+   //     applying any necessary transformations such as: eliminating boundary
+   //     conditions, applying conforming constraints for non-conforming AMR,
+   //     static condensation, etc.
+
+   if (pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+   a.Assemble(0);
+
+   OperatorPtr A;
+   Vector B, X;
+   a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
+
+   cout << "Size of linear system: " << A->Height() << endl;
+
+   ConstantCoefficient absomeg(pow2(omega) * epsilon);
+   RestrictedCoefficient restr_absomeg(absomeg,attr);
+
+   ParBilinearForm prec(fespace);
+   prec.AddDomainIntegrator(new CurlCurlIntegrator(restr_muinv));
+   prec.AddDomainIntegrator(new VectorFEMassIntegrator(restr_absomeg));
+
+   PMLDiagMatrixCoefficient pml_c1_abs(cdim,detJ_inv_JT_J_abs, pml);
+   ScalarVectorProductCoefficient c1_abs(muinv,pml_c1_abs);
+   VectorRestrictedCoefficient restr_c1_abs(c1_abs,attrPML);
+
+   PMLDiagMatrixCoefficient pml_c2_abs(dim, detJ_JT_J_inv_abs,pml);
+   ScalarVectorProductCoefficient c2_abs(absomeg,pml_c2_abs);
+   VectorRestrictedCoefficient restr_c2_abs(c2_abs,attrPML);
+
+   prec.AddDomainIntegrator(new CurlCurlIntegrator(restr_c1_abs));
+   prec.AddDomainIntegrator(new VectorFEMassIntegrator(restr_c2_abs));
+
+   prec.Assemble();
+
+   // 11. Define and apply a GMRES solver for AU=B with a block diagonal
+   //     preconditioner based on the Gauss-Seidel or Jacobi sparse smoother.
+   Array<int> offsets(3);
+   offsets[0] = 0;
+   offsets[1] = fespace->GetTrueVSize();
+   offsets[2] = fespace->GetTrueVSize();
+   offsets.PartialSum();
+
+   std::unique_ptr<Operator> pc_r;
+   std::unique_ptr<Operator> pc_i;
+   real_t s = (conv == ComplexOperator::HERMITIAN) ? -1_r : 1_r;
+
+   OperatorPtr PCOpAh;
+   prec.SetDiagonalPolicy(mfem::Operator::DIAG_ONE);
+   prec.FormSystemMatrix(ess_tdof_list, PCOpAh);
+
+
+   // Gauss-Seidel Smoother
```

Comment:
Is Gauss-Seidel an accurate description for this block diagonal AMG preconditioner?